### ISSUE TITLE
Suppress warning for generated code

### ISSUE THREAD

**Issue by vsfeedback** [dotnet/roslyn#41171]
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/content/idea/887015/suppress-warning-for-generated-code.html)._

---
<p>Once upon a time it was possible to suppress <code>CS1591:Missing XML comment for publicly visible type or member</code> for generated code via that properties page (Solution Explorer -&gt; Properties -&gt; Code Analysis Settings).</p> <p>How can this be established now for .net core solutions?</p> <p><br>I have tried the following, but that does not seem to work, as it is a compiler warning and not a code analysis warning?<br> <code>[assembly: SuppressMessage(&quot;Compiler&quot;, &quot;CS1591&quot;, Justification=&quot;Generated Code&quot;, Scope=&quot;namespaceanddescendants&quot;, Target=&quot;MyNamespaceContainingGeneratedCode&quot;)]</code></p> <p>The only option I see at the moment, is to move all generated code to a separate project and just set <code>&lt;noWarn&gt;1591&lt;/noWarn&gt;</code> in the project file. A viable option, but it falls short if you want to add your own logic to some of the generated <code>partial</code> classes.</p> <p><a href="https://github.com/MicrosoftDocs/visualstudio-docs/issues/3871">https://github.com/MicrosoftDocs/visualstudio-docs/issues/3871<br></a>Related to: <a href="https://github.com/MicrosoftDocs/visualstudio-docs/issues/3230">#3230</a></p><blockquote>
<p>Solution Explorer -&gt; Properties -&gt; Code Analysis Settings:<br>NOTE: This property page has been deprecated and will be removed in a future product release.</p></blockquote>

---
#### Visual Studio Feedback System on 1/16/2020, 10:45 PM: 

Thank you for taking the time to provide your suggestion.  We will do some preliminary checks to make sure we can proceed further.  We'll provide an update once the issue has been triaged by the product team.

**Comment by mavasani** [comment]
This looks like a compiler issue. User wants to suppress specific compiler warnings in generated code, but as far as I know the compiler itself never supported a project/compilation wide setting for it in past. I believe user has 2 options, if that helps:

1.	Apply `#pragma warning disable CS1591` at the top of each generated code file. I presume user may not want to do this as this involves editing generated code, which they likely don’t hand edit.
2.	Use .editorconfig based suppression for all the relevant file patterns. Something like below in an .editorconfig at root of their project/solution:

```
[*.designer.cs]
dotnet_diagnostic.CS1591.severity = none
```

**Comment by mikadumont** [comment]
Will reply back with the two possible solutions.

**Comment by boukeversteegh** [comment]
I'd be curious to know if there are any options that will suppress these warnings while compiling, and do not involve patching generated code

**Comment by Siphonophora** [comment]
Hi, I have `dotnet_diagnostic.CS1591.severity = none` in my editor config file, and that seems to suppress all CS1591 except for those coming from generated code. 

I'm not 100% sure this is the same as what's described above.

**Comment by boukeversteegh** [comment]
> Hi, I have `dotnet_diagnostic.CS1591.severity = none` in my editor config file, and that seems to suppress all CS1591 except for those coming from generated code.
> 
> I'm not 100% sure this is the same as what's described above.

I think that is the inverse of what is meant here. The idea is that generated code may contain problems that you cannot fix, and you would like to suppress the errors _only_ for the generated files, not the other way around.

**Comment by Siphonophora** [comment]
@boukeversteegh yeah, you are right. https://github.com/dotnet/roslyn/issues/54119 is what I am seeing

**Comment by earloc** [comment]
I have a similar situation. While generating code my `source-generator` might [emit warnings](https://github.com/dotnet/roslyn/blob/main/docs/features/source-generators.cookbook.md#issue-diagnostics).
However, users might not be able to suppress such warnings, which might be very annoying especially if [TreatWarningsAsErrors](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-options/errors-warnings#treatwarningsaserrors) is enabled.

It would be a nice addition if users could override the `DiagnosticSeverity` without authors of `source-generators` (like me ;)) need to implement workarounds to support this scenario. See https://github.com/earloc/TypealizR/pull/35
where I needed to implement some kind of "polyfill" to enable this kind of customization.

**Comment by hankovich** [comment]
> Will reply back with the two possible solutions.

@mikadumont we've been waiting for three long years...

**Comment by sharwell** [comment]
CS1591 can be suppressed using the [`DiagnosticSuppressor`](https://learn.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.diagnostics.diagnosticsuppressor?view=roslyn-dotnet-4.6.0) API. 

Here's an example of a diagnostic suppressor for a different scenario:
https://github.com/dotnet/roslyn-analyzers/blob/48e7c8f02b8e38e3bba0411923a904f25e9bf1a0/src/Roslyn.Diagnostics.Analyzers/Core/RelaxTestNamingSuppressor.cs

**Comment by hankovich** [comment]
@sharwell how to supress CS0618 (member is obsolete). 

Such section in editorconfig

```
[*.{g,designer}.cs]

# CS0618: Member is obsolete
dotnet_diagnostic.CS0618.severity = none
```

does not work for some reason.

**Comment by sharwell** [comment]
> how to supress CS0618 (member is obsolete).

@hankovich this seems to be a different situation. Can you create a new issue with a code sample for the generator which is resulting in CS0618?

**Comment by hankovich** [comment]
> Can you create a new issue with a code sample for the generator which is resulting in CS0618?

@sharwell [sure](https://github.com/dotnet/roslyn/issues/69298). For some reason in my repro project it produces CS0612, but the issue still exists.

**Comment by sveinungf** [comment]
One workaround that seems to be working for me is to suppress warnings globally in a `.globalconfig` file, and then re-enable them in an `.editorconfig` file. Then I'm no longer receiving warnings from source generated code, but I still get warnings from my hand-written code. Both files are placed in the same folder as the solution file.

My `.globalconfig` file:
```
is_global = true

dotnet_diagnostic.RS0030.severity = none
```

From my `.editorconfig` file:
```
[*.cs]
dotnet_diagnostic.RS0030.severity = warning
```

**Comment by Anakael** [comment]
I have a problem with generated source that does not put `GeneratedCode` attribute. If it does - no CS1591 is raised.
Example:
https://github.com/sveinungf/spreadcheetah/issues/138

In my `.editorconfig` I have:
```
[*.cs]
dotnet_diagnostic.CS1591.severity = none
```
and
```
[*.g.cs]
generated_code = true
```

but it does not affect CS1591 warning on generated code.

**Comment by DennisLindnerSLS** [comment]
I can confirm that this issue unfortunately still exists in VS2026.

In our project, we enforce strict coding guidelines and require XML documentation for all developers using the following project settings:
```
<TreatWarningsAsErrors>True</TreatWarningsAsErrors>
<EnforceCodeStyleInBuild>True</EnforceCodeStyleInBuild>
<GenerateDocumentationFile>true</GenerateDocumentationFile>
```

Unfortunately, we receive the `Missing XML comment for publicly visible type or member` warning for files that are added via NuGet references. Attempts to suppress this warning by modifying our `.editorconfig` for these specific files:
```
dotnet_diagnostic.CS1591.severity = none
dotnet_diagnostic.IDE0005.severity = none
dotnet_analyzer_diagnostic.severity = none
generated_code = true
```
have not resolved the issue. As a result, we are forced to add these warnings to our global suppression list, which essentially defeats the purpose of enforcing XML documentation as part of our build process.

**Comment by CyrusNajmabadi** [comment]
> Unfortunately, we receive the Missing XML comment for publicly visible type or member warning for files that are added via NuGet references. 

These files are violating your standards. Downstream consumers have no way to know which public types you expose were generated vs not. So this is correctly blocking things since you have stated it's an error to ship undocumented types. 

> As a result, we are forced to add these warnings to our global suppression list, which essentially defeats the purpose of enforcing XML documentation as part of our build process.

That seems appropriate.  Either you have this enforcement, in which case blocking the violation is correct. Or you want to let it actually ship, in which case you do need to suppress.

Now, if you want a hybrid, where you still ship some undocumented types, you can use a diagnostic suppressor to suppress things. That's a core scenario these were designed for.
